finetuning-llama-7b-on-l4/README.md

```diff
@@ -0,0 +1,381 @@
+
+The nodepool has been created and is scaled down to 0 nodes. So you are not paying for any GPUs until you start launching K8s Pods that request GPUs.
+
+## Run a K8s job to finetune Llama 2 7B
+
+Finetuning requires a base model and a dataset. For this post, the [dell-research-harvard/AmericanStories](https://huggingface.co/datasets/dell-research-harvard/AmericanStories) dataset will be used to finetune the [Llama 2 7B](https://huggingface.co/meta-llama/Llama-2-7b-hf) base model. GCS will be used for storing the base model. GKE with GCSFuse is used to transparently save the finetuned model to GCS. This provides a cost efficient way to store and serve the model and only pay for the storage used by the model.
+
+
+### Configuring GCS and required permissions
+
+Create a GCS bucket to store our models:
+```bash
+gcloud storage buckets create gs://${BUCKET_NAME}
+```
+
+The model loading Job will write to GCS. So let’s create a Google Service Account that has read and write permissions to the GCS bucket. Then create a K8s Service Account named `l4-demo` that is able to use the Google Service Account.
+
+To do this, first create a new Google Service Account:
+```bash
+gcloud iam service-accounts create l4-demo
+```
+
+Assign the required GCS permissions to the Google Service Account:
+```bash
+gcloud storage buckets add-iam-policy-binding gs://${BUCKET_NAME} \
+  --member="serviceAccount:${SERVICE_ACCOUNT}" --role=roles/storage.admin
+```
+
+Allow the K8s Service Account `l4-demo` in the `default` namespace to use the Google Service Account:
+```bash
+gcloud iam service-accounts add-iam-policy-binding ${SERVICE_ACCOUNT} \
+  --role roles/iam.workloadIdentityUser \
+  --member "serviceAccount:${PROJECT_ID}.svc.id.goog[default/l4-demo]"
+```
+
+Create a new K8s Service Account:
+```bash
+kubectl apply -f - << EOF
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: l4-demo
+  namespace: default
+  annotations:
+    iam.gke.io/gcp-service-account: l4-demo@$PROJECT_ID.iam.gserviceaccount.com
+EOF
+```
+
+Hugging face requires authentication to download the[ Llama 2 7B HF](https://huggingface.co/meta-llama/Llama-2-7b-hf) model, which means an access token is required to download the model.
+
+You can get your access token from [huggingface.com > Settings > Access Tokens](https://huggingface.co/settings/tokens). Afterwards, set your HuggingFace token as an environment variable:
+```bash
+export HF_TOKEN=<paste-your-own-token>
+```
+
+Create a Secret to store your HuggingFace token which will be used by the K8s job:
+```bash
+kubectl create secret generic l4-demo --from-literal="HF_TOKEN=$HF_TOKEN"
+```
+
+Run a K8s Job to download the the Llama 2 7B model to the bucket which is mounted to the K8s Pod:
+```bash
+kubectl apply -f - << EOF
```

Comment:
I wrote this as a guide people can run through without having to clone repos and switch contexts, mostly because it's originally written for a blog post.

I think now that I'm moving it to this repo it does make sense to have finetune.py, Dockerfile and job manifests to be normal files. The first step would then be cloning the repo. I tried to get away with least amount of effort to convert my blog post in Gdoc to this repo. Your comments are all very valid and I think I should just do it.